pubtools/_content_gateway/push_base.py

```diff
@@ -480,33 +534,140 @@ def process_file(self, file_item):
-                self.file_mapping[
-                    (product_name, product_code, version_name, download_url)] = file_id
-                LOG.info("New file created with file_id:- %s" % file_id)
-                return
-
-            LOG.info("Found file record with file_id:- %s" % file_id)
-            LOG.info("Updating the file metadata")
+        #  If CGW mapping doesn't have the record it will try to get file_id from metadata if provided
+        #  For the create operation it will not get any file_id neither from the CGW
+        #  nor from metadata. The file_id would be None
+        #  To update the downloadURL the file_id must be provided in the metadata
+        file_id = file_id if file_id else file_item.get('metadata').get('id')
+
+        del file_item.get('metadata')['productName'], \
+            file_item.get('metadata')['productCode'], \
+            file_item.get('metadata')['productVersionName']
+        file_item.get('metadata')['productVersionId'] = version_id
+
+        if file_item.get('state') == 'create':
+            LOG.info("Creating file entry for the given downloadURL: %s ", download_url)
+            file_id = self.cgw_client.create_file(product_id, version_id, file_item.get('metadata'))
+            self.file_mapping[(product_name, product_code, version_name, download_url)] = file_id
+            LOG.info("New file created with file_id:- %s \n" % file_id)
+
+            # adding operational product_id and file_id for rollback operation
+            file_item['product_id'] = product_id
+            file_item['file_id'] = file_id
+            # keeping a copy of data that has been used to create the cgw data
+            self.completed_operations.append(file_item)
+            return
+
+        if file_item.get('state') == 'update' and file_id:
+            LOG.info("Updating the file record \n")
             file_item.get('metadata')['id'] = file_id
             self.cgw_client.update_file(product_id, version_id, file_item.get('metadata'))
-        elif file_item.get('state') == 'absent' and file_id:
+
+            # adding operational product_id for rollback operation
+            file_item['product_id'] = product_id
+            # keeping a copy of data that has been used to update the cgw data
+            self.completed_operations.append(file_item)
+
+        elif file_item.get("state") == "delete" and file_id:
             LOG.info("Deleting existing file records for file_id:- %s" % file_id)
             self.cgw_client.delete_file(product_id, version_id, file_id)
-            LOG.info("File record deleted!")
+
+            #  removing the deleted file record from the completed operations
+            #  this is to avoid the conflicts during the rollback and make_visible operations
+            for record in self.completed_operations:
+                if record['type'] == 'file' and record['file_id'] == file_id:
+                    self.completed_operations.remove(record)
```

Comment:
Will deleted data not be rolled back?

Reply:
Yes, deleted data cannot be rolled back.
Rollback operation will only work for create and update operations.